app/src/main/java/pm/gnosis/heimdall/ui/safe/create/ConfirmSafeRecoveryPhraseActivity.kt

```diff
@@ -0,0 +1,157 @@
+
+    // Default composite disposable gets cleared on onStop while clicks need to survive the lifetime of the activity
+    private var wordClickDisposables = CompositeDisposable()
+
+    @Inject
+    lateinit var adapter: ConfirmSafeRecoveryPhraseAdapter
+
+    override fun onCreate(savedInstanceState: Bundle?) {
+        super.onCreate(savedInstanceState)
+
+        val encryptedMnemonic = intent.getStringExtra(EXTRA_ENCRYPTED_MNEMONIC)
+        val chromeExtensionAddress = intent.getStringExtra(EXTRA_CHROME_EXTENSION_ADDRESS).asEthereumAddress()
+        if (encryptedMnemonic == null || chromeExtensionAddress == null) {
+            finish(); return
+        }
+
+        disposables += viewModel.setup(encryptedMnemonic, chromeExtensionAddress)
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeBy(onSuccess = ::onSuccessfulSetup, onError = { finish() })
+
+        layout_confirm_safe_recovery_phrase_selected_words.apply {
+            setHasFixedSize(true)
+            layoutManager = this@ConfirmSafeRecoveryPhraseActivity.layoutManager
+            adapter = this@ConfirmSafeRecoveryPhraseActivity.adapter
+        }
+    }
+
+    override fun onStart() {
+        super.onStart()
+        disposables += layout_confirm_safe_recovery_phrase_finish.clicks()
+            .flatMapSingle {
+                viewModel.createSafe()
+                    .doOnSubscribe {
+                        layout_confirm_safe_recovery_phrase_progress_bar.visibility = View.VISIBLE
+                        bottomBarEnabled(false)
+                    }
+            }
+            .observeOn(AndroidSchedulers.mainThread())
+            .subscribeForResult(onNext = ::onSafeCreated, onError = ::onSafeCreationError)
+
+        disposables += layout_confirm_safe_recovery_phrase_back.clicks()
+            .subscribeBy(onNext = { finish() }, onError = Timber::e)
+    }
+
+    private fun onIsCorrectSequence(isCorrectSequence: Boolean) {
+        bottomBarEnabled(isCorrectSequence)
+    }
+
+    private fun onIsCorrectSequenceError(throwable: Throwable) {
+        Timber.e(throwable)
+        bottomBarEnabled(false)
+    }
+
+    private fun bottomBarEnabled(enable: Boolean) {
+        layout_confirm_safe_recovery_phrase_finish.isEnabled = enable
+        layout_confirm_safe_recovery_phrase_bottom_bar.setBackgroundColor(getColorCompat(if (enable) R.color.azure else R.color.bluey_grey))
+    }
+
+    private fun onSafeCreated(txHash: BigInteger) {
+        startActivity(SafeMainActivity.createIntent(this, txHash))
+        toast("Safe created successfully")
+    }
+
+    private fun onSafeCreationError(throwable: Throwable) {
+        Timber.e(throwable)
+        layout_confirm_safe_recovery_phrase_progress_bar.visibility = View.INVISIBLE
+        bottomBarEnabled(true)
+        toast("Something went wrong")
+    }
+
+    private fun onSuccessfulSetup(words: List<String>) {
+        words.forEachIndexed { index, word ->
+            when {
```

Comment:
super dynamic xD

Reply:
Indeed. Was thinking of a recycler view with a grid layout for this one. Was mostly because of time constraints.